server/src/repositories/asset.repository.ts

```diff
@@ -27,155 +43,169 @@ import {
   WithProperty,
   WithoutProperty,
 } from 'src/interfaces/asset.interface';
-import { AssetSearchOptions, SearchExploreItem } from 'src/interfaces/search.interface';
-import { searchAssetBuilder } from 'src/utils/database';
-import { Paginated, PaginationOptions, paginate, paginatedBuilder } from 'src/utils/pagination';
-import {
-  Brackets,
-  FindOptionsOrder,
-  FindOptionsRelations,
-  FindOptionsSelect,
-  FindOptionsWhere,
-  In,
-  IsNull,
-  MoreThan,
-  Not,
-  Repository,
-} from 'typeorm';
-
-const truncateMap: Record<TimeBucketSize, string> = {
-  [TimeBucketSize.DAY]: 'day',
-  [TimeBucketSize.MONTH]: 'month',
-};
-
-const dateTrunc = (options: TimeBucketOptions) =>
-  `(date_trunc('${
-    truncateMap[options.size]
-  }', (asset."localDateTime" at time zone 'UTC')) at time zone 'UTC')::timestamptz`;
+import { MapMarker, MapMarkerSearchOptions } from 'src/interfaces/map.interface';
+import { AssetSearchOptions, SearchExploreItem, SearchExploreItemSet } from 'src/interfaces/search.interface';
+import { anyUuid, asUuid, introspectTables, mapUpsertColumns } from 'src/utils/database';
+import { Paginated, PaginationOptions, paginationHelper } from 'src/utils/pagination';
+import { DataSource } from 'typeorm';
 
 @Injectable()
 export class AssetRepository implements IAssetRepository {
   constructor(
-    @InjectRepository(AssetEntity) private repository: Repository<AssetEntity>,
-    @InjectRepository(AssetFileEntity) private fileRepository: Repository<AssetFileEntity>,
-    @InjectRepository(ExifEntity) private exifRepository: Repository<ExifEntity>,
-    @InjectRepository(AssetJobStatusEntity) private jobStatusRepository: Repository<AssetJobStatusEntity>,
-  ) {}
-
-  async upsertExif(exif: Partial<ExifEntity>): Promise<void> {
-    await this.exifRepository.upsert(exif, { conflictPaths: ['assetId'] });
+    @InjectKysely() private db: Kysely<DB>,
+    @InjectDataSource() private dataSource: DataSource,
+  ) {
+    introspectTables(this.dataSource, 'exif', 'asset_job_status', 'asset_files');
+  }
+
+  async upsertExif(exif: Insertable<Exif>): Promise<void> {
+    const value = { ...exif, assetId: asUuid(exif.assetId) };
+    await this.db
+      .insertInto('exif')
+      .values(value)
+      .onConflict((oc) =>
+        oc.columns(EXIF_CONFLICT_KEYS).doUpdateSet(() => mapUpsertColumns('exif', value, EXIF_CONFLICT_KEYS)),
+      )
+      .execute();
   }
 
-  async upsertJobStatus(...jobStatus: Partial<AssetJobStatusEntity>[]): Promise<void> {
-    await this.jobStatusRepository.upsert(jobStatus, { conflictPaths: ['assetId'] });
-  }
+  async upsertJobStatus(...jobStatus: Insertable<AssetJobStatus>[]): Promise<void> {
+    if (jobStatus.length === 0) {
+      return;
+    }
```

Comment:
There are three reasons for it:
1. Kysely doesn't handle it unfortunately
2. The code checks the first element of the array to see which fields are being upserted
3. It resolves in one tick instead of two by avoiding the `await` keyword